public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx

```diff
@@ -1,35 +1,109 @@
 
 import { getNavModel } from '../../core/selectors/navModel';
 import { StoreState } from '../../types';
 
 interface ServiceAccountCreatePageProps {
   navModel: NavModel;
 }
-interface ServiceAccountDTO {
-  name: string;
-}
 
 const createServiceAccount = async (sa: ServiceAccountDTO) => getBackendSrv().post('/api/serviceaccounts/', sa);
+const updateServiceAccount = async (id: number, sa: ServiceAccountDTO) =>
+  getBackendSrv().patch(`/api/serviceaccounts/${id}`, sa);
 
 const ServiceAccountCreatePage: React.FC<ServiceAccountCreatePageProps> = ({ navModel }) => {
+  const [roleOptions, setRoleOptions] = useState<Role[]>([]);
+  const [builtinRoles, setBuiltinRoles] = useState<{ [key: string]: Role[] }>({});
+  const [pendingRoles, setPendingRoles] = useState<Role[]>([]);
+
+  const currentOrgId = contextSrv.user.orgId;
+  const [serviceAccount, setServiceAccount] = useState<ServiceAccountDTO>({
+    id: 0,
+    orgId: contextSrv.user.orgId,
+    role: OrgRole.Viewer,
+    tokens: 0,
+    name: '',
+    login: '',
+    isDisabled: false,
+    createdAt: '',
+    teams: [],
+  });
+
+  useEffect(() => {
+    async function fetchOptions() {
+      try {
+        if (contextSrv.hasPermission(AccessControlAction.ActionRolesList)) {
+          let options = await fetchRoleOptions(currentOrgId);
+          setRoleOptions(options);
+        }
+
+        if (contextSrv.hasPermission(AccessControlAction.ActionBuiltinRolesList)) {
+          const builtInRoles = await fetchBuiltinRoles(currentOrgId);
+          setBuiltinRoles(builtInRoles);
+        }
+      } catch (e) {
+        console.error('Error loading options', e);
+      }
+    }
+    if (contextSrv.licensedAccessControlEnabled()) {
+      fetchOptions();
+    }
+  }, [currentOrgId]);
+
   const history = useHistory();
 
   const onSubmit = useCallback(
     async (data: ServiceAccountDTO) => {
-      await createServiceAccount(data);
+      data.role = serviceAccount.role;
+      const response = await createServiceAccount(data);
+      try {
+        const newAccount: ServiceAccountCreateApiResponse = {
+          avatarUrl: response.avatarUrl,
+          id: response.id,
+          isDisabled: response.isDisabled,
+          login: response.login,
+          name: response.name,
+          orgId: response.orgId,
+          role: response.role,
+          tokens: response.tokens,
+        };
+        const updateResponse = await updateServiceAccount(response.id, data);
+        // TODO: error handling
+        console.log(updateResponse);
+        const updateRolesResponse = await updateUserRoles(pendingRoles, newAccount.id, newAccount.orgId);
+        // TODO: error handling
+        console.log(updateRolesResponse);
+      } catch (e) {
+        console.error(e);
+      }
       history.push('/org/serviceaccounts/');
     },
-    [history]
+    [history, serviceAccount.role, pendingRoles]
```

Comment:
aha, so you create a service account and within the request you use the update api to update the roles. 🤔 

Interesting. I was working on making the role apart of the creation api. But that does not need to happen. This just creates two queries instead.

nice approach